src/modalities/trainer.py

```diff
@@ -106,19 +110,39 @@ def _train_batch(
             scheduler (LRScheduler): The learning rate scheduler.
             loss_fun (Loss): The loss function used for training.
             micro_batch_id (int): The ID of the micro batch.
+            scheduled_pipeline (Optional[Pipeline], optional): In case of pipeline parallelism, this is used to
+                operate the model. Defaults to None.
 
         Returns:
             tuple[bool, int, torch.Tensor, Optional[torch.Tensor], Optional[torch.Tensor]]:
                 A tuple containing the following:
                     - step_performed (bool): Indicates whether a training step was performed.
                     - num_train_steps_done (int): The number of training steps done.
-                    - loss (torch.Tensor): The computed loss.
+                    - loss (Optional[torch.Tensor]): The computed loss.
+                        None, if a non-last stage was processes in pipeline parallelism.
                     - gradient_norm_score (Optional[torch.Tensor]): The gradient norm score,
                         if a training step was performed otherwise return None.
         """
-        result_batch = model_predict_batch(model=model, batch=batch)
-        loss = loss_fun(result_batch)
-        (loss / self.gradient_acc_steps).backward()
+        if scheduled_pipeline is not None:
+            pp_schedule = scheduled_pipeline.pp_schedule
+            # Pipeline Parallel forward / backward inside step() call
+            # with self.train_context(optional_context_parallel_ctx):
+            targets, losses = (
+                (batch.targets[loss_fun.target_key].contiguous(), [])
+                if scheduled_pipeline.is_last_pp_stage
+                else (None, None)
+            )
+
+            if scheduled_pipeline.is_first_pp_stage:
+                pp_schedule.step(batch.samples[model.sample_key].contiguous(), target=targets, losses=losses)
+            else:
+                pp_schedule.step(target=targets, losses=losses)
+            loss = torch.mean(torch.stack(losses)).to(losses[0].device) if scheduled_pipeline.is_last_pp_stage else None
+        else:
+            # else continue with loss calculation
```

Comment:
This code is very specific to pipeline parallelism. I would prefer if that would be abstracted away. However I also have no clear picture yet, how to do that. We should discuss this.